### PR TITLE
[DTensor] Coalesce collectives for multi-arg redistribution

### LINKED ISSUE #170367: DTensor redistributes on foreach_X operators are inefficient
### 🐛 Describe the bug

When you have a `foreach_X` operator this is typically because you need to run a lot of teeny tiny kernels and it is profitable to do a single kernel launch for all of them. In some situations, in DTensor, it may be necessary to redistribute all of the inputs e.g. from Replicate -> Partial, and the current implementation is bad because it will issue a non-foreach kernel for every input tensor in this case.

Typical stack looks like

```
torch/distributed/tensor/_dispatch.py(233): _dispatch_get_local_results_slow_path
torch/distributed/tensor/_dispatch.py(436): redistribute_local_args
torch/distributed/tensor/_redistribute.py(789): redistribute_local_tensor
torch/distributed/tensor/_ops/_math_ops.py(91): _partition_value
```

### Versions

main

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx

### PR DESCRIPTION
**Pull request by stmcgovern**
Fixes https://github.com/pytorch/pytorch/issues/170367

When multiple op arguments need the same redistribution (same mesh, same src/dst placements), batch them into a single coalesced collective call instead of issuing N separate NCCL kernel launches. This is most impactful for foreach/fused ops (e.g. _foreach_add_, _fused_adam_) during optimizer steps with many parameters, but applies to any multi-arg op.

Supports all_reduce, reduce_scatter, and allgather coalescing (all_to_all is not coalesced as there is no coalesced variant in funcol). Flattened mesh transforms (multi-dim mesh optimization) are also supported. Falls back to per-tensor redistribution for unsupported cases (e.g. _MaskPartial, multi-step transforms, subclassed placement types).

### CODE REVIEW

File: test/distributed/tensor/test_tensor_ops.py [hunk 1]
```diff
@@ -1344,6 +1344,281 @@ def test_foreach_mixed_dtensor_and_tensor(self):
...
+            self.assertEqual(s.full_tensor(), e)
+
+    @skip_if_lt_x_gpu(2)
+    @with_comms
+    def test_coalesced_matches_per_tensor_redistribute(self):
+        """Coalesced and per-tensor redistribution must produce identical results.
+
+        This is the key invariant: if the per-tensor paths in placement_types
+        (Shard._reduce_shard_tensor, Shard._to_replicate_tensor) evolve,
+        this test catches any divergence from the coalesced paths.
```
Comment (wconstab) [inline test/distributed/tensor/test_tensor_ops.py]: i didn't understand this comment. can you elaborate?

File: torch/distributed/tensor/_dispatch.py [hunk 2]
```diff
@@ -598,6 +589,60 @@ def redistribute_local_args(
...
+                    reshard_arg_spec,
+                )
+        else:
+            resharded = redistribute_local_tensors(
+                [(t, src, dst) for _, t, src, dst in reshard_requests]
+            )
+            for (idx, _, arg_spec, reshard_arg_spec), result in zip(
+                reshard_requests, resharded
+            ):
+                ExplicitRedistributionContext.observe_redistribution(
```
Comment (wconstab) [inline torch/distributed/tensor/_dispatch.py]: wondering if we can find a way to collapse some of this code. can we observe the redistribution in the same loop where we append the redist requests?

That could have 2 nice properties:
1) no dupe observe_redist calls
2) if we trigger a redistribution error, we stop on the first one which is the same behavior as we would have had in the pre-merge case, and we stop before actually doing the redistribution

also, for debug mode, unclear if we want to log the pre- or post- optimized thing. it seems like we should be logging what acutally happened in debug mode, so we have observability into the optimization. i'd expect some way of distinguishing from the debug-mode logs that the optimization happened and the single redistribution was issued for the individual tensors.

File: test/distributed/tensor/test_tensor_ops.py (hunk 1 above)
Comment (stmcgovern) [reply test/distributed/tensor/test_tensor_ops.py]: clarified this to explain that the test is just meant to enforce that the per-tensor path and coalesced path match.

File: torch/distributed/tensor/_dispatch.py [hunk 3]
```diff
@@ -598,6 +602,41 @@ def redistribute_local_args(
...
+                new_local_args[idx] = redistribute_local_tensor(
+                    local_tensor,
+                    arg_spec,
+                    reshard_arg_spec,
+                )
+        else:
+            items = [(t, src, dst) for _, t, src, dst in reshard_requests]
+            resharded = redistribute_local_tensors(items)
+            for (idx, _, arg_spec, reshard_arg_spec), result in zip(
+                reshard_requests, resharded
```
Comment (Skylion007) [inline torch/distributed/tensor/_dispatch.py]: Pass strict kwarg here to

File: torch/distributed/tensor/_redistribute.py [hunk 4]
```diff
@@ -1456,6 +1456,229 @@ def _gen_transform_infos(
...
+    groups: defaultdict[tuple, list[int]] = defaultdict(list)
+    for i, (_, src, dst) in enumerate(items):
+        groups[(id(src.mesh), src.placements, dst.placements)].append(i)
+
+    # Dict insertion order ensures all ranks issue collectives in the same order.
+    for indices in groups.values():
+        if len(indices) > 1:
+            coalesced = _try_coalesced_redistribute([items[i] for i in indices])
+            if coalesced is not None:
+                for idx, result in zip(indices, coalesced):
```
Comment (Skylion007) [inline torch/distributed/tensor/_redistribute.py]: Same here

File: torch/distributed/tensor/_redistribute.py [hunk 5]
```diff
@@ -1456,6 +1456,229 @@ def _gen_transform_infos(
...
+    for tensor, _, _ in items:
+        if guard_or_true(tensor.size(shard_dim) % num_chunks != 0):
+            scattered_list, pad_sizes = shard_spec._split_tensor(
+                tensor, num_chunks, with_padding=True, contiguous=True
+            )
+            padded_tensors.append(torch.cat(scattered_list, dim=shard_dim))
+            pad_sizes_list.append(pad_sizes)
+        else:
+            padded_tensors.append(
+                tensor.contiguous() if not tensor.is_contiguous() else tensor
```
Comment (Skylion007) [inline torch/distributed/tensor/_redistribute.py]: Isn't contiguous a noop if it's already contiguous anyway?

Comment (stmcgovern) [reply torch/distributed/tensor/_redistribute.py]: yep thanks for the improvements

**Comment by stmcgovern** [comment]
cc @ezyang

**Comment by ezyang** [comment]
@claude evaluate the potential blast radius of the PR. How risky is it, what people are likely to be affected?

**Comment by ezyang** [comment]
This is pretty long and I am concerned about hard to predict downstream implications. How much do you care about landing this? (I did file the bug, but due to changes in strategy these days I'm not super gung ho about big DTensor improvements). If you do care about getting this in, we need to actually do some benchmarks.

**Review by ezyang — CHANGES_REQUESTED** [review]
https://github.com/pytorch/pytorch/wiki/Pull-request-review-conventions

**Comment by stmcgovern** [comment]
> How much do you care about landing this? (I did file the bug, but due to changes in strategy these days I'm not super gung ho about big DTensor improvements).

It's not high priority at the moment.

**Comment by stmcgovern** [comment]
not a priority at the moment.
